src/cdk-experimental/menu/menu-aim.ts

```diff
@@ -0,0 +1,271 @@
+import {fromEvent, Subject} from 'rxjs';
+import {takeUntil, filter} from 'rxjs/operators';
+import {PointerFocusTracker, FocusableElement} from './pointer-focus-tracker';
+import {Menu} from './menu-interface';
+import {throwMissingPointerFocusTracker, throwMissingMenuReference} from './menu-errors';
+
+/**
+ * MenuAim is responsible for determining if a sibling menuitem's menu should be closed when a
+ * Toggler item is hovered into. It is up to the hovered in item to call the MenuAim service in
+ * order to determine if it may perform its close actions.
+ */
+export interface MenuAim {
+  /** Set the mouse manager tracking the menu's menu items. */
+  withMouseManager(pointerTracker: PointerFocusTracker<FocusableElement & Toggler>): this;
+
+  /** Set the menu for which this tracks mouse movements in. */
+  withMenu(menu: Menu): this;
+
+  /**
+   * Calls the `doToggle` callback when it is deemed that the user is not moving towards
+   * the submenu.
+   * @param doToggle the function called when the user is not moving towards the submenu.
+   */
+  toggle(doToggle: () => void): void;
+}
+
+/** Injection token used for an implementation of MenuAim. */
+export const MENU_AIM = new InjectionToken<MenuAim>('cdk-menu-aim');
+
+/** Capture every nth mouse move event. */
+const MOUSE_MOVE_SAMPLE_FREQUENCY = 3;
+
+/** The number of mouse move events to track. */
+const NUM_POINTS = 5;
+
+/**
+ * How long to wait before closing a sibling menu if a user stops short of the submenu they were
+ * predicted to go into.
+ */
+const CLOSE_DELAY = 300;
+
+/**
+ * An element which when hovered over may perform closing actions on the open submenu and
+ * potentially open its own menu.
+ */
+export interface Toggler {
+  getMenu(): Menu | undefined;
+}
+
+/** Calculate the slope between point a and b. */
+function slope(a: Point, b: Point) {
+  return (b.y - a.y) / (b.x - a.x);
+}
+
+/** Calculate the y intercept for the given point and slope. */
+function yIntercept(point: Point, _slope: number) {
+  return point.y - _slope * point.x;
+}
+
+/** Represents a coordinate of mouse travel. */
+type Point = {x: number; y: number};
+
+/**
+ * The bounding top and bottom points within the opened submenu, and the left, middle and right x
+ * coordinates.
+ */
+type SubmenuPoints = {bottom: number; top: number; x: number; midX: number; endX: number};
+
+/** Whether the given line defined by the slope and y intercept falls within the submenu. */
+function isWithinSubmenu(submenuPoints: SubmenuPoints, _slope: number, yintercept: number) {
+  return (
+    (_slope * submenuPoints.x + yintercept > submenuPoints.top &&
+      _slope * submenuPoints.x + yintercept < submenuPoints.bottom) ||
+    (_slope * submenuPoints.midX + yintercept > submenuPoints.top &&
+      _slope * submenuPoints.midX + yintercept < submenuPoints.bottom) ||
+    (_slope * submenuPoints.endX + yintercept > submenuPoints.top &&
+      _slope * submenuPoints.endX + yintercept < submenuPoints.bottom)
```

Comment:
```suggestion
  // y = mx + b
  // x = (y - b) / m
  const m = slope;
  const b = yintercept;
  const {left, right, top, bottom} = submenuPoints;

  // Check for intersection with each edge of the submenu (left, right, top, bottom)
  // by fixing one coordinate to that edge's coordinate (either x or y) and checking if the
  // other coordinate is within bounds.
  return (
      (m * left + b >= top && m * left + b <= bottom) ||
      (m * right + b >= top && m * right + b <= bottom) ||
      ((top - b) / m >= left && (top - b) / m <= right) ||
      ((bottom - b) / m >= left && (bottom - b) / m <= right)
  );
```
I realized it's more correct to do this while also being basically the same amount fo code because you don't have to grab the midpoint any more. This also eliminates the `SubmenuPoints` type so you can just use `ClientRect`